### PR TITLE
Mldsa functional tests

### LINKED ISSUE #641: Run mldsa-native functional tests in our CI
The goal of this issue is to run a small selection of functional tests of mldsa-native as part of `all.sh`. We only want coverage of functional testing on the platforms that we target.

What this means in terms of test coverage:

* Here, we're only testing mldsa-native, the way we _build_ it (compilation options, target platform, compiler flags), but not the way we _integrate_ it (we're not calling any of our integration code).
* We rely on the test coverage in mldsa-native for algorithmic tests, including edge cases.
* Maintainer tests (lint, portability, proofs, etc.) are out of scope here.

TBD: what about constant-time testing?

I'm still investigating, but at this stage I think what we need to do is to run [`scripts/tests`](https://github.com/pq-code-package/mldsa-native/blob/606e8ddf7be8a7284ad415fca464e4695fb19647/scripts/tests) with the right options, similar to the [`functest` action](https://github.com/pq-code-package/mldsa-native/blob/main/.github/actions/functest/action.yml).

Prerequisites: https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/628

### PR DESCRIPTION
**Pull request by bjwtaylor**
## Description

Add mldsa functional tests to all.sh, resolves https://github.com/Mbed-TLS/TF-PSA-Crypto/issues/641

## PR checklist

- [ ] **changelog** not required because:  No functional changes.
- [ ] **framework PR** not required
- [ ] **TF-PSA-Crypto development PR** provided #HERE
- [ ] **TF-PSA-Crypto 1.1 PR** not required because:  No backports
- [ ] **mbedtls development PR** not required because: No changes
- [ ] **mbedtls 4.1 PR** not required because: No changes
- [ ] **mbedtls 3.6 PR**not required because: No changes
- **tests**  provided

### CODE REVIEW

File: tests/scripts/components-pqcp.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+################################################################
+
+support_test_pqcp_mldsa_native_upstream_all () {
+    python3 -c 'import sys; exit(1 if sys.version_info < (3, 9) else 0)'
+}
+
+# This condition deliberately runs when all.sh sources this file. It hides the
+# component from both `--list-all-components` and explicit component requests on
+# runners whose Python is too old for mldsa-native's generated-test checks.
+if support_test_pqcp_mldsa_native_upstream_all; then
```
Comment (valeriosetti) [inline tests/scripts/components-pqcp.sh]: Sure I'm missing something here, but is this `if` statement required?
I read the comment above that explains it, but as far as I remember the machinery that instruments our testing already does this. In support of my comment I don't see any other test `component_xxx` with an associated `support_xxx` with the `if` check.

**Review by valeriosetti — CHANGES_REQUESTED** [review]

File: tests/scripts/components-pqcp.sh (hunk 1 above)
Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: So, it's been a while since I looked at this. However I think because the mldsa-native test cases require certain versions of python we need this to verify we are running that specific version and only run when we are.

Comment (valeriosetti) [reply tests/scripts/components-pqcp.sh]: Yes, this is similar to what other `support_xxx` function do. IIRC the CI initially get a list of which `component_xxx` should be run on each specific ubuntu version, so there should be no need for extra logic.
However I might miss something so let's wait for the 2nd reviewer to confirm/deny my hypothesis.

Comment (gilles-peskine-arm) [reply tests/scripts/components-pqcp.sh]: As Valerio says, this is exactly what `support_xxx` does. The `component_xxx` function should be defined unconditionally. It will only run on a platform where `support_xxx` is true. See for example `support_tf_psa_crypto_check_committed_generated_files` with the same code for the same reason.

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: OK, I'lll update it.

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: I've pushed the changes, lets see what the CI thinks.

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: CI is passing, so I think this one is good for re-review.

File: tests/scripts/components-pqcp.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+#### PQCP Testing
+################################################################
+
+support_test_pqcp_mldsa_native_upstream_all () {
+    python3 -c 'import sys; exit(1 if sys.version_info < (3, 9) else 0)'
+}
+
+# This condition deliberately runs when all.sh sources this file. It hides the
+# component from both `--list-all-components` and explicit component requests on
+# runners whose Python is too old for mldsa-native's generated-test checks.
```
Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh]: Obsolete comment, please remove it.

File: tests/scripts/components-pqcp.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+component_test_pqcp_mldsa_native_upstream_all () {
+    msg "test: mldsa-native upstream all --opt ALL --examples"
+     cd drivers/pqcp/mldsa-native
+    ./scripts/tests all --opt ALL --examples
```
Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh]: Do we care about examples?

Also, `--examples` is (currently) enabled by default.

Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh]: Should we pass `--stack` (off by default)? We do care about stack usage in some projects, so having the upstream information could be useful.

Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh]: We do want the alloc tests. `--alloc` is currently enabled by default, but we may want to make it explicit.

Comment (gilles-peskine-arm) [reply tests/scripts/components-pqcp.sh]: Yes, we do care about examples. That's how mldsa-native tests itself with an external SHAKE. There's also an example that tests `MLD_CONFIG_REDUCE_RAM`.

File: tests/scripts/components-pqcp.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+################################################################
+
+support_test_pqcp_mldsa_native_upstream_all () {
+    python3 -c 'import sys; exit(1 if sys.version_info < (3, 9) else 0)'
+}
+
+# This condition deliberately runs when all.sh sources this file. It hides the
+# component from both `--list-all-components` and explicit component requests on
+# runners whose Python is too old for mldsa-native's generated-test checks.
+component_test_pqcp_mldsa_native_upstream_all () {
```
Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh:16]: This component tests mldsa-native in its default configuration, and with a set of variations, including an external SHAKE.

This set of variations does not cover one variant that we care about, which is using a custom memory allocator. So we should also test that. (Done via `.github/actions/config-variations/action.yml` in the mldsa-native CI.) We should (in the same component, given that it only takes a few minutes) run `scripts/tests all` with `--cflags='-DMLD_CONFIG_FILE "configs/custom_heap_alloc_config.h"'` (may require additional quoting) (plus the other options).

Comment (gilles-peskine-arm) [inline tests/scripts/components-pqcp.sh:16]: Should we test other targets (architectures, compilers)? They already do upstream, but not with the same compilers, and anyway our goal here is to ensure that the exact version that we integrate is tested.

I think that we don't need to test additional targets that are already tested upstream. We should test IAR, which is not tested upstream, but that hinges on having IAR on the CI, which we currently don't. So that's a future action item.

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've reviewed based on my analysis of the mldsa-native CI and on how we use it.

File: tests/scripts/components-pqcp.sh (hunk 3 above)
Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: OK, it's probably best to leave the --examples in then in case they disable it's default in the future? If so are you happy for this to be closed?

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: So, it is probably useful as it will monitor the stack usage which we are probably concerned about. I think it will increase CI time and will need valgrind, though if we don't have this already we can presumably add it. I've added the option let's see if the CI passes, or if it requires additional dependencies in the environment.

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh]: now added.

Comment (gilles-peskine-arm) [reply tests/scripts/components-pqcp.sh]: We do have valgrind on the CI. Let's see how much load it adds.

File: tests/scripts/components-pqcp.sh (hunk 4 above)
Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh:16]: Now added

Comment (bjwtaylor) [reply tests/scripts/components-pqcp.sh:16]: Yeah, I think a lot of this is already tested upstream. So we should focus on the things that really matter to us. So I'll leave these out, let me know if you really want them though.

File: tests/scripts/components-pqcp.sh (hunk 3 above)
Comment (gilles-peskine-arm) [reply tests/scripts/components-pqcp.sh]: The stack analysis takes about 3 second per build in https://ci.trustedfirmware.org/job/mbed-tls-tf-psa-crypto-multibranch/job/PR-780-head/10/cloudbees-pipeline-explorer/?filter=1299 . This is significantly less than the build (~30s). So no performance problem here.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM

**Review by valeriosetti — APPROVED** [review]
